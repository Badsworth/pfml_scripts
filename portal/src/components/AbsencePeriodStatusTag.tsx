import {
  AbsencePeriodRequestDecision,
  AbsencePeriodRequestDecisionEnum,
} from "../models/AbsencePeriod";
import Tag, { TagProps } from "./core/Tag";
import React from "react";
import findKeyByValue from "../utils/findKeyByValue";
import { useTranslation } from "../locales/i18n";

interface AbsencePeriodStatusTagProps {
  request_decision?: AbsencePeriodRequestDecisionEnum;
}

const StateMap: {
  [status in keyof typeof AbsencePeriodRequestDecision]: TagProps["state"];
} = {
  approved: "success",
  cancelled: "inactive",
  denied: "error",
  inReview: "warning",
  pending: "warning",
  projected: "warning",
  withdrawn: "inactive",
  voided: "inactive",
} as const;

const AbsencePeriodStatusTag = ({
  request_decision,
}: AbsencePeriodStatusTagProps) => {
  const { t } = useTranslation();
  const requestDecisionKey = findKeyByValue(
    AbsencePeriodRequestDecision,
    request_decision
  );

  if (!requestDecisionKey) return null;

  return (
    <Tag
      label={t("components.absencePeriodStatusTag.label", {
        context: requestDecisionKey,
      })}
      state={StateMap[requestDecisionKey]}
    />
  );
};

export default AbsencePeriodStatusTag;
