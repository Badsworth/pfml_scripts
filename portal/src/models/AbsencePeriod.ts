import { LeaveReasonType } from "./LeaveReason";
import { groupBy } from "lodash";

export type AbsencePeriodRequestDecision =
  | "Approved"
  | "Cancelled"
  | "Denied"
  | "In Review"
  | "Pending"
  | "Completed"
  // Projected is (at least) used for Bonding Leave periods created automatically
  // for periods with Birth Disability as the reason qualifier
  | "Projected"
  | "Voided"
  | "Withdrawn";

export type AbsencePeriodTypes =
  | "Continuous"
  | "Intermittent"
  | "Reduced Schedule";

export class AbsencePeriod {
  absence_period_end_date: string;
  absence_period_start_date: string;
  fineos_leave_request_id: string | null = null;
  period_type: AbsencePeriodTypes;
  reason: LeaveReasonType;
  reason_qualifier_one = "";
  reason_qualifier_two = "";
  request_decision: AbsencePeriodRequestDecision;

  constructor(attrs: Partial<AbsencePeriod> = {}) {
    Object.assign(this, attrs);
  }

  static groupByReason(absence_periods: AbsencePeriod[]): {
    [reason: string]: AbsencePeriod[];
  } {
    return groupBy(absence_periods, "reason");
  }

  /**
   * @returns periods sorted newest to oldest (by start date)
   */
  static sortNewToOld(absence_periods: AbsencePeriod[]): AbsencePeriod[] {
    const periods = absence_periods.slice(); // avoids mutating the original array
    return periods.sort((a, b) => {
      return a.absence_period_start_date > b.absence_period_start_date ? -1 : 1;
    });
  }
}
