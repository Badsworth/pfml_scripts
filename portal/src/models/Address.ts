/* eslint sort-keys: ["error", "asc"] */
import BaseModel from "./BaseModel";

class Address extends BaseModel {
  // @ts-expect-error ts-migrate(2416) FIXME: Property 'defaults' in type 'Address' is not assig... Remove this comment to see the full error message
  get defaults() {
    return {
      city: null,
      line_1: null,
      line_2: null,
      state: null,
      zip: null,
    };
  }
}

export default Address;
