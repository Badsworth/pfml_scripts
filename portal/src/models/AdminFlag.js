import BaseModel from "./BaseModel";

class FlagModel extends BaseModel {
  get defaults() {
    return {
      flag_id: null,
      name: null,
      enabled: false,
      options: {},
      start: null,
      end: null,
    };
  }
}

export default FlagModel;
