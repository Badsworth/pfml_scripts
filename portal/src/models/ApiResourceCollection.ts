export interface ApiResource {
  // API resources are expected to be objects. We expect at least an ID property.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

/**
 * Readonly generic class representing a collection of a specific data model. Methods
 * return new instances of the collection rather than modifying the original collection.
 * @example const c = new ApiResourceCollection<Document>("document_id", [{ document_id: "123", name: "example" }]);
 *          c.setItem({ document_id: "456", name: "example2" });
 */
class ApiResourceCollection<TApiResource extends ApiResource> {
  idKey: keyof TApiResource;
  private map: Map<string, TApiResource>;

  constructor(idKey: keyof TApiResource, items: TApiResource[] = []) {
    this.idKey = idKey;
    this.map = new Map(
      items.map((item) => {
        return [item[idKey], item];
      })
    );
  }

  private get _mutableMap() {
    return new Map(this.map);
  }

  private _arrayFromMap(map: Map<string, TApiResource>): TApiResource[] {
    return Array.from(map.values());
  }

  get items(): TApiResource[] {
    return this._arrayFromMap(this.map);
  }

  get isEmpty(): boolean {
    return this.map.size === 0;
  }

  getItem(itemId: string): TApiResource | undefined {
    return this.map.get(itemId);
  }

  /**
   * Add or update an item in the collection. Returns a new collection with the
   * item included. Does not modify the original collection.
   */
  setItem(item: TApiResource): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap.set(item[this.idKey], item);
    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }

  /**
   * Add or update items in the collection. Returns a new collection with the
   * items included. Does not modify the original collection.
   */
  setItems(items: TApiResource[] = []): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap;
    items.forEach((item) => {
      updatedMap.set(item[this.idKey], item);
    });
    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }

  /**
   * Removes an item from the collection. Returns a new collection with the item removed.
   * Does not modify the original collection.
   */
  removeItem(itemId: string): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap;
    updatedMap.delete(itemId);

    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }
}

export default ApiResourceCollection;
