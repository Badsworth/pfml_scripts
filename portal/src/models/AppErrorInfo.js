import BaseModel from "./BaseModel";
import uniqueId from "lodash/uniqueId";

/**
 * Provides a consistent interface for creating error messages displayed to the user.
 * @property {string} [field] - Path of the field the error is associated with
 * @property {string} key - Unique key used for React `key` prop (https://reactjs.org/docs/lists-and-keys.html)
 * @property {string} message - Internationalized message displayed to the user (like `Error#message`)
 * @property {string} [name] - Name of the error (like `Error#name`)
 */
class AppErrorInfo extends BaseModel {
  get defaults() {
    return {
      key: uniqueId("AppErrorInfo"),
      message: null,
      name: null,
      field: null,
    };
  }
}

export default AppErrorInfo;
