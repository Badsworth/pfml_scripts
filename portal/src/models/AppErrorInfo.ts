import { uniqueId } from "lodash";

/**
 * Provides a consistent interface for creating error messages displayed to the user.
 * @property [field] - Path of the field the error is associated with
 * @property key - Unique key used for React `key` prop (https://reactjs.org/docs/lists-and-keys.html)
 * @property message - Internationalized message displayed to the user (like `Error#message`)
 * @property [name] - Name of the error (like `Error#name`)
 * @property [meta] - Additional error data, like the application_id associated with it
 * @property [rule] - Name of validation issue rule (e.g "min_leave_periods", "conditional", etc)
 * @property [type] - Name of validation issue type (e.g "required", "pattern", "date", etc)
 */
class AppErrorInfo {
  key: string = uniqueId("AppErrorInfo");
  message?: string | JSX.Element;
  name?: string;
  field?: string;
  meta?: Record<string, unknown>;
  rule?: string;
  type?: string;

  constructor(attrs: Partial<AppErrorInfo>) {
    Object.assign(this, attrs);
  }
}

export default AppErrorInfo;
