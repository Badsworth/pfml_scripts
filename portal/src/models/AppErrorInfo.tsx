import React from "react";
import { uniqueId } from "lodash";

/**
 * Provides a consistent interface for creating error messages displayed to the user.
 * @property [field] - Path of the field the error is associated with
 * @property key - Unique key used for React `key` prop (https://reactjs.org/docs/lists-and-keys.html)
 * @property message - Internationalized message displayed to the user (like `Error#message`)
 * @property [name] - Name of the error (like `Error#name`)
 * @property [meta] - Additional error data, like the application_id associated with it
 * @property [rule] - Name of validation issue rule (e.g "min_leave_periods", "conditional", etc)
 * @property [type] - Name of validation issue type (e.g "required", "pattern", "date", etc)
 */
class AppErrorInfo {
  key: string = uniqueId("AppErrorInfo");
  message?: string | JSX.Element;
  name?: string;
  field?: string;
  meta?: { [key: string]: unknown };
  rule?: string;
  type?: string;

  constructor(attrs: Partial<AppErrorInfo>) {
    Object.assign(this, attrs);
  }

  /**
   * Get a human readable error message for the given field, if an associated error exists.
   * @param errors - All errors, including those not associated with the given field
   * @param field - The field to get the error message(s) for
   */
  static fieldErrorMessage(errors: AppErrorInfo[], field: string) {
    if (!errors.length) return null;

    const fieldErrors = errors.filter((error) => error.field === field);

    if (!fieldErrors.length) {
      return null;
    }

    // Combine all errors for this field into a single message to be rendered
    return fieldErrors.map((error, i) => (
      <React.Fragment key={error.key}>
        {error.message}
        {i < fieldErrors.length - 1 && " "}
      </React.Fragment>
    ));
  }
}

export default AppErrorInfo;
