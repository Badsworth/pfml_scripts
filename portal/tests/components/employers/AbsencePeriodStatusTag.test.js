import { AbsencePeriodRequestDecision } from "../../../src/models/AbsencePeriod";
import AbsencePeriodStatusTag from "../../../src/components/AbsencePeriodStatusTag";
import React from "react";
import { render } from "@testing-library/react";

describe("AbsencePeriodStatusTag", () => {
  it.each(Object.values(AbsencePeriodRequestDecision))(
    "renders tag for %s",
    (decision) => {
      const { container } = render(
        <AbsencePeriodStatusTag request_decision={decision} />
      );
      expect(container.firstChild).toMatchSnapshot();
    }
  );

  it("returns null when the decision is undefined", () => {
    const { container } = render(<AbsencePeriodStatusTag />);

    expect(container).toBeEmptyDOMElement();
  });

  it("accepts custom classes", () => {
    const { container } = render(
      <AbsencePeriodStatusTag
        request_decision={AbsencePeriodRequestDecision.approved}
        className="width-12"
      />
    );

    expect(container.firstChild).toHaveClass("width-12");
  });
});
