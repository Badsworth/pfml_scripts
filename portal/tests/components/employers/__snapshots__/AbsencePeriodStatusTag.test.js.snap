// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbsencePeriodStatusTag renders tag for Approved 1`] = `
<span
  class="usa-tag display-inline-block text-bold text-middle text-center text-no-wrap text-success-dark bg-success-lighter"
>
  Approved
</span>
`;

exports[`AbsencePeriodStatusTag renders tag for Cancelled 1`] = `
<span
  class="usa-tag display-inline-block text-bold text-middle text-center text-no-wrap text-base-darker bg-base-lightest"
>
  Cancelled
</span>
`;

exports[`AbsencePeriodStatusTag renders tag for Denied 1`] = `
<span
  class="usa-tag display-inline-block text-bold text-middle text-center text-no-wrap text-error-darker bg-error-lighter"
>
  Denied
</span>
`;

exports[`AbsencePeriodStatusTag renders tag for In Review 1`] = `
<span
  class="usa-tag display-inline-block text-bold text-middle text-center text-no-wrap text-base-darkest bg-warning-lighter"
>
  Pending
</span>
`;

exports[`AbsencePeriodStatusTag renders tag for Pending 1`] = `
<span
  class="usa-tag display-inline-block text-bold text-middle text-center text-no-wrap text-base-darkest bg-warning-lighter"
>
  Pending
</span>
`;

exports[`AbsencePeriodStatusTag renders tag for Projected 1`] = `
<span
  class="usa-tag display-inline-block text-bold text-middle text-center text-no-wrap text-base-darkest bg-warning-lighter"
>
  Pending
</span>
`;

exports[`AbsencePeriodStatusTag renders tag for Voided 1`] = `
<span
  class="usa-tag display-inline-block text-bold text-middle text-center text-no-wrap text-base-darker bg-base-lightest"
>
  Cancelled
</span>
`;

exports[`AbsencePeriodStatusTag renders tag for Withdrawn 1`] = `
<span
  class="usa-tag display-inline-block text-bold text-middle text-center text-no-wrap text-base-darker bg-base-lightest"
>
  Withdrawn
</span>
`;
